src/models/state/wallet/wallet_state.rs

```diff
@@ -338,40 +339,46 @@ impl WalletState {
             StrongUtxoKey,
             (MsMembershipProof, u64),
         > = HashMap::default();
-        for (i, monitored_utxo) in monitored_utxos.iter() {
-            let utxo_digest = Hash::hash(&monitored_utxo.utxo);
 
-            match monitored_utxo
-                .get_membership_proof_for_block(new_block.kernel.header.prev_block_digest)
-            {
-                Some(ms_mp) => {
-                    debug!("Found valid mp for UTXO");
-                    let replacement_success = valid_membership_proofs_and_own_utxo_count.insert(
-                        StrongUtxoKey::new(utxo_digest, ms_mp.auth_path_aocl.leaf_index),
-                        (ms_mp, i),
-                    );
-                    assert!(
-                        replacement_success.is_none(),
-                        "Strong key must be unique in wallet DB"
-                    );
-                }
-                None => {
-                    // Was MUTXO marked as abandoned? Then this is fine. Otherwise, log a warning.
-                    // TODO: If MUTXO was spent, maybe we also don't want to maintain it?
-                    if monitored_utxo.abandoned_at.is_some() {
-                        debug!("Monitored UTXO with digest {utxo_digest} was marked as abandoned. Skipping.");
-                    } else {
-                        let confirmed_in_block_info = match monitored_utxo.confirmed_in_block {
-                            Some(mutxo_received_in_block) => format!(
-                                "UTXO was received at block height {}.",
-                                mutxo_received_in_block.2
-                            ),
-                            None => String::from("No info about when UTXO was confirmed."),
-                        };
-                        warn!(
-                        "Unable to find valid membership proof for UTXO with digest {utxo_digest}. {confirmed_in_block_info} Current block height is {}", new_block.kernel.header.height
-                    );
-                        // panic!("Unable to find valid membership proof.");
+        {
+            let stream = monitored_utxos.stream().await;
+            pin_mut!(stream); // needed for iteration
+
+            while let Some((i, monitored_utxo)) = stream.next().await {
```

Comment:
Same as above: Aren't we reading from the DB `monitored_utxos.len()` many times here, whereas we used to just be reading once? Note that monitored UTXOs is unlikely to be more than a few thousand instances and can thus safely fit into RAM.

Reply:
The new logic is equivalent to the old code.  Both are reading line-by-line.   Same general comments apply as above.

Detail:

The old code for the for loop was: 

```
   for (i, monitored_utxo) in monitored_utxos.iter() {
       ...
   }
```

this is equivalent:

```
    while let Some((i, monitored_utxo)) = stream.next().await {
        ...
    |
```


With regards to len(), it is called exactly once, in StorageVecStream::stream(), which looks like:

```
    async fn stream<'a>(&'a self) -> impl Stream<Item = (Index, T)> + 'a
    where
        T: 'a,
    {
        self.stream_many(0..self.len().await).await
    }
```

and stream_many looks like:

```
    async fn stream_many<'a>(
        &'a self,
        indices: impl IntoIterator<Item = Index> + 'a,
    ) -> impl Stream<Item = (Index, T)> + 'a
    where
        T: 'a,
    {
        stream! {
            for i in indices.into_iter() {
                yield (i, self.get(i).await)
            }
        }
    }
```

With regards to memory usage and concurrency, see my comment above.